src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -204,6 +205,51 @@ protected Solution SetCurrentSolution(Solution solution)
             }
         }
 
+        /// <summary>
+        /// Applies specified transformation to <see cref="CurrentSolution"/> and updates <see cref="CurrentSolution"/> to the new value.
+        /// </summary>
+        /// <param name="transformation">Solution transformation.</param>
+        /// <param name="oldSolution">Previous instance of the solution.</param>
+        /// <param name="newSolution">Transformed instance of the solution with updated <see cref="Solution.WorkspaceVersion"/>.</param>
+        /// <returns>True if <see cref="CurrentSolution"/> was set to the transformed solution, false if the transformation did not change the solution.</returns>
+        protected bool SetCurrentSolution(Func<Solution, Solution> transformation, out Solution oldSolution, out Solution newSolution)
+        {
+            if (transformation == null)
+            {
+                throw new ArgumentNullException(nameof(transformation));
+            }
+
+            var currentSolution = Volatile.Read(ref _latestSolution);
+
+            while (true)
+            {
+                var transformedSolution = transformation(currentSolution);
+                if (transformedSolution == currentSolution)
+                {
+                    oldSolution = newSolution = currentSolution;
+                    return false;
+                }
+
+                newSolution = transformedSolution.WithNewWorkspace(this, currentSolution.WorkspaceVersion + 1);
```

Comment:
There's a part of me that wishes this was a local variable and we didn't assign to the out until inside the if; the subtle bit being that if the assignment *fails* any code change here could accidentally not remember to reassign newSolution and leak something odd.

Reply:
Makes sense.